packages/fxa-auth-server/scripts/stripe-products-and-plans-to-firestore-documents/stripe-products-and-plans-converter.ts

```diff
@@ -0,0 +1,451 @@
+          stripeObject.metadata,
+          metadataKeyPrefix
+        );
+        if (value.length > 0) {
+          uiContentConfig[key] = value;
+        }
+        continue;
+      }
+      value =
+        stripeObject.metadata[key] ?? stripeObject.metadata[`product:${key}`];
+      if (value) {
+        // @ts-ignore We need to be able to assign values to all keys
+        uiContentConfig[key] = value;
+      }
+    }
+    return uiContentConfig;
+  }
+
+  /**
+   * Convert Stripe Product or Plan URL metadata keys with various prefixes
+   * and suffixes to a UrlConfig object.
+   */
+  urlMetadataToUrlConfig(
+    stripeObject: Stripe.Product | Stripe.Plan
+  ): UrlConfig {
+    const urlConfig: UrlConfig = {};
+    if (!stripeObject.metadata) {
+      return urlConfig;
+    }
+    for (const key of UrlConfigKeys) {
+      const value =
+        stripeObject.metadata[`${key}URL`] ??
+        stripeObject.metadata[`${key}Link`] ??
+        stripeObject.metadata[`product:${key}URL`];
+      if (value) {
+        urlConfig[key] = value;
+      }
+    }
+    return urlConfig;
+  }
+
+  /**
+   * Convert a Stripe Product to a ProductConfig object with an empty
+   * `locales` key.
+   */
+  stripeProductToProductConfig(product: Stripe.Product): ProductConfig {
+    // Firestore cannot serialize class instances
+    const productConfig: any = {};
+
+    // BaseConfig
+    productConfig.active = product.active;
+    productConfig.capabilities =
+      this.capabilitiesMetadataToCapabilityConfig(product);
+    // `locales` will be populated once we iterate through the Stripe Plans
+    productConfig.locales = {};
+    productConfig.styles = this.stylesMetadataToStyleConfig(product);
+    productConfig.support = this.supportMetadataToSupportConfig(product);
+    productConfig.uiContent = this.uiContentMetadataToUiContentConfig(product);
+    productConfig.urls = this.urlMetadataToUrlConfig(product);
+
+    // Extended by ProductConfig
+    productConfig.stripeProductId = product.id;
+    const { productSet, promotionCodes } = product.metadata;
+    if (productSet) {
+      productConfig.productSet = productSet;
+    }
+    if (promotionCodes) {
+      productConfig.promotionCodes = commaSeparatedListToArray(promotionCodes);
+    }
+    return productConfig;
+  }
+
+  /**
+   * Infer a locale (language only or language and region) from a Stripe Plan.
+   * TODO: #12053: Improve heuristics
+   */
+  findLocaleStringFromStripePlan(plan: Stripe.Plan): null | string {
+    // Try to extract a locale from the plan nickname
+    const { nickname } = plan;
+    let locale = nickname
+      ?.split(' ')
+      .filter((w) => this.supportedLanguages.includes(w.toLowerCase()));
+    if (locale && locale.length > 0) {
+      return locale[0];
+    }
+    return null;
+  }
+
+  /**
+   * Extract localized data from a Stripe Plan and convert it to
+   * ProductConfig.locales
+   */
+  stripePlanLocalesToProductConfigLocales(
```

Comment:
@chenba : I'm not currently handling metadata keys that have a `{locale}` appended to the end, as I don’t see any RPs making use of this in stage or production, and I wanted to minimize scope creep. What do you think?

I.e. maybe we should just remove that from the docs and recommend the one plan per billing period per locale approach for everyone until we have a better solution in the code for localizing these strings? Else maybe this can be a follow-up as well.

Reply:
Yeah, it's likely fine.  We've been discouraging that way of doing l10n on the metadata because we have a bug with metadata l10n.  🙃